src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -2062,6 +2059,83 @@ void ChannelData::DestroyResolverAndLbPolicyLocked() {
   }
 }
 
+void ChannelData::UpdateStateAndPickerLocked(
+    grpc_connectivity_state state, const absl::Status& status,
+    const char* reason,
+    std::unique_ptr<LoadBalancingPolicy::SubchannelPicker> picker) {
+  // Clean the control plane when entering IDLE.
+  if (picker == nullptr || state == GRPC_CHANNEL_SHUTDOWN) {
+    health_check_service_name_.reset();
+    saved_service_config_.reset();
+    saved_config_selector_.reset();
+  }
+  // Update connectivity state.
+  state_tracker_.SetState(state, status, reason);
+  if (channelz_node_ != nullptr) {
+    channelz_node_->SetConnectivityState(state);
+    channelz_node_->AddTraceEvent(
+        channelz::ChannelTrace::Severity::Info,
+        grpc_slice_from_static_string(
+            channelz::ChannelNode::GetChannelConnectivityStateChangeString(
+                state)));
+  }
+  // Grab data plane lock to do subchannel updates and update the picker.
+  //
+  // Note that we want to minimize the work done while holding the data
+  // plane lock, to keep the critical section small.  So, for all of the
+  // objects that we might wind up unreffing here, we actually hold onto
+  // the refs until after we release the lock, and then unref them at
+  // that point.  This includes the following:
+  // - refs to subchannel wrappers in the keys of pending_subchannel_updates_
+  // - ref stored in retry_throttle_data_
+  // - ref stored in service_config_
+  // - ref stored in config_selector_
+  // - ownership of the existing picker in picker_
+  RefCountedPtr<ServerRetryThrottleData> retry_throttle_data_to_unref;
+  RefCountedPtr<ServiceConfig> service_config_to_unref;
+  RefCountedPtr<ConfigSelector> config_selector_to_unref;
+  {
+    MutexLock lock(&data_plane_mu_);
```

Comment:
Would like to understand the non-inversion property here (order of locks, whatever)

Reply:
No code every holds `data_plane_mu_` and `resolution_mu_` at the same time, so there's no need to define ordering.